docs/internal/concurrency/modelimpl/When.cs

```diff
@@ -0,0 +1,217 @@
+/**
+ *   Common part of a cown that is independent of the data the cown is storing.
+ *
+ *   Just contains a pointer to the last behaviour's request for this cown.
+ */
+class CownBase : StableOrder
+{
+    // Points to the end of the queue of requests for this cown.
+    // If it is null, then the cown is not currently in use by 
+    // any behaviour.
+    volatile internal Request? last = null;
+}
+
+class Behaviour
+{
+    // The body of the behaviour.
+    Action thunk;
+    // How many requests are outstanding for the behaviour.
+    int count;
+    // The set of requests for this behaviour.
+    // This is used to release the cowns to the subseqeuent behaviours.
+    Request[] requests;
+
+    internal Behaviour(Action t, CownBase[] cowns)
+    {
+        thunk = t;
+        // We add an additional count, so that the 2PL is finished
+        // before we start running the thunk. Without this, the calls to
+        // Release at the end of the thunk could race with the calls to 
+        // FinishEnqueue in the 2PL.
+        count = cowns.Count() + 1;
+
+        Array.Sort(cowns);
+        requests = new Request[cowns.Length];
+        for (int i = 0; i < cowns.Length; i++)
+        {
+            requests[i] = new Request(this, cowns[i]);
+        }
+
+        // Complete first phase of 2PL enqueuing on all cowns.
+        for (int i = 0; i < cowns.Length; i++)
+        {
+            requests[i].StartEnqueue();
+        }
+
+        // Complete second phase of 2PL enqueuing on all cowns.
+        for (int i = 0; i < cowns.Length; i++)
+        {
+            requests[i].FinishEnqueue();
+        }
+
+        // Resolve the additional request.
```

Comment:
The way the code and comment is written, it is a bit sneaky that this also runs (schedules) the thunk.

Reply:
ah sorry it doesn't!

Reply:
OK, hope my confusion was useful. Your comment is excellent!